src/main.rs

```diff
@@ -612,6 +638,86 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         tokio::time::sleep(std::time::Duration::from_secs(1)).await;
     }
 
+    if num_rpc_nodes > 0 {
+        let rpc_node = cluster_images.rpc()?;
+        let mut rpc_nodes = vec![];
+        // Create and deploy rpc secrets
+        for rpc_index in 0..num_rpc_nodes {
+            let rpc_secret = kub_controller.create_rpc_secret(rpc_index, &config_directory)?;
+            rpc_node.set_secret(rpc_secret);
+            kub_controller.deploy_secret(rpc_node.secret()).await?;
+            info!("Deployed RPC node {rpc_index} Secret");
+
+            let identity_path =
+                config_directory.join(format!("rpc-node-identity-{rpc_index}.json"));
+            let rpc_keypair =
+                read_keypair_file(identity_path).expect("Failed to read rpc-node keypair file");
+
+            rpc_node.add_label(
+                "rpc-node/name",
+                &format!("rpc-node-{rpc_index}"),
+                LabelType::Service,
+            );
+
+            rpc_node.add_label(
+                "rpc-node/type",
+                rpc_node.validator_type().to_string(),
+                LabelType::Info,
+            );
+
+            rpc_node.add_label(
+                "rpc-node/identity",
+                rpc_keypair.pubkey().to_string(),
+                LabelType::Info,
+            );
+
+            rpc_node.add_label(
+                "load-balancer/name",
+                "load-balancer-selector",
+                LabelType::Service,
+            );
+
+            let replica_set = kub_controller.create_rpc_replica_set(
+                rpc_node.image(),
+                rpc_node.secret().metadata.name.clone(),
+                &rpc_node.all_labels(),
+                rpc_index,
+            )?;
+            rpc_node.set_replica_set(replica_set);
+
+            kub_controller
+                .deploy_replicas_set(rpc_node.replica_set())
+                .await?;
+            info!("rpc node replica set ({rpc_index}) deployed successfully");
+
+            let rpc_service = kub_controller.create_service(
+                &format!("rpc-node-selector-{rpc_index}"),
+                rpc_node.service_labels(),
+            );
+            kub_controller.deploy_service(&rpc_service).await?;
+            info!("rpc node service ({rpc_index}) deployed successfully");
+
+            rpc_nodes.push(rpc_node.replica_set_name().clone());
+        }
+
+        // wait for at least one rpc node to deploy
+        loop {
+            let mut ready = false;
+            for rpc_name in &rpc_nodes {
+                if kub_controller.is_replica_set_ready(rpc_name).await? {
+                    ready = true;
+                    break;
+                }
+            }
+            if ready {
+                break;
+            }
+
+            info!("no rpc nodes ready yet");
+            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+        }
```

Comment:
micro-nit: you could use a labled loop for this to be a bit cleaner, ie:

```
        'outer: loop {
            for rpc_name in &rpc_nodes {
                if kub_controller.is_replica_set_ready(rpc_name).await? {
                    break 'outer;
                }
            }

            info!("no rpc nodes ready yet");
            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
        }
```

More info at https://doc.rust-lang.org/rust-by-example/flow_control/loop/nested.html

Reply:
ohhh lets goooo. this is much better. thank you!